core/thread.c

```diff
@@ -55,6 +55,47 @@ const char *thread_getname(kernel_pid_t pid)
 #endif
 }
 
+void thread_zombify(void)
+{
+    if (irq_is_in()) {
+        return;
+    }
+
+    unsigned state = irq_disable();
+    sched_set_status((thread_t *)sched_active_thread, STATUS_ZOMBIE);
+    irq_restore(state);
+    thread_yield_higher();
+
+    sched_task_exit();
+}
+
+int thread_kill_zombie(kernel_pid_t pid)
+{
+    DEBUG("thread_kill: Trying to kill PID %" PRIkernel_pid "...\n", pid);
+    unsigned state = irq_disable();
+
+    thread_t *thread = (thread_t *) thread_get(pid);
+
+    if (!thread) {
+        DEBUG("thread_kill: Thread does not exist!\n");
+    }
+    else if (thread->status == STATUS_ZOMBIE) {
+        DEBUG("thread_kill: Thread is a zombie.\n");
+
+        sched_threads[pid] = NULL;
+        sched_num_threads--;
+        sched_set_status(thread, STATUS_STOPPED); 
+
+        irq_restore(state);
```

Comment:
how about:

```
unsigned state = ...
int res = (int)STATUS_NOT_FOUND;

...
if () {
...
}
else if (thread->status == STATUS_ZOMBIE) {
  ...
  res = 0;
}
else {
 ...
}
irq_restore(state);
return res;
```

- saves one "irq_restore()"
- changes return to "0 on success, <0 on error"

Reply:
> changes return to "0 on success, <0 on error"

scratch that, seems like other `thread_*` functions also return 1 on success.

Reply:
>     * saves one "irq_restore()"

we can do this as followup.